### PR TITLE
Allow multiple trusted networks with differing proxy header behavior (raises IpSpoofAttackError)

### PR DESCRIPTION
**Pull request by edslocomb**
request.remote_ip() should not raise ActionDispatch::RemoteIp::IpSpoofAttackError when a request has passed through proxies for two or more private (or otherwise trusted) networks, where one proxy has set the nonstandard HTTP_CLIENT_IP header, and another has set the nonstandard HTTP_X_FORWARDED_FOR header.

As it stands, e.g.:
1. A request originates at 192.168.1.103, and passes through a proxy which sets HTTP_CLIENT_IP to that address.
2. The first proxy then sends the request out from 10.1.1.232, to a second proxy
3. The second proxy prepends this new address to HTTP_X_FORWARDED_FOR, and then sends the request to a Rails server
4. The Rails server sees that both HTTP_CLIENT_IP and HTTP_X_FORWARDED_FOR have been set, so it compares the two addresses in the two headers.
5. Since the addresses are different, Rails raises an ActionDispatch::RemoteIp::IpSpoofAttackError

The fix is to filter trusted IPs before making this comparison (tests included).

### CODE REVIEW

File: actionpack/lib/action_dispatch/middleware/remote_ip.rb [hunk 1]
```diff
@@ -131,18 +131,26 @@ def initialize(env, middleware)
...
-        # Set by the Rack web server, this is a single value.
+        # We assume these things about the IP headers:
+        #
+        #   - X-Forwarded-For will be a list of IPs, one per proxy, or blank
+        #   - Client-Ip is propagated from the outermost proxy, or is blank
+        #   - either of the above could be a CSV list and/or repeated headers
+        #       that were concatenated.
+        #   - REMOTE_ADDR will be the IP that made the request to Rack
+
         remote_addr = ips_from('REMOTE_ADDR').last
```
Comment (edslocomb) [inline actionpack/lib/action_dispatch/middleware/remote_ip.rb]: As of 4.2 the regex has been removed, I don't think there are any remaining perfomance considerations for this PR.

**Comment by edslocomb** [comment]
Can we have another look at this?  

The performance questions have been cleared up, leaving only the problematic behavior described (and exposed by the tests included in this PR).

**Comment by maclover7** [comment]
Please rebase.

**Comment by edslocomb** [comment]
rebased.

**Pull request closed without merging**